lib/algolia.ex

```diff
@@ -568,4 +568,41 @@ defmodule Algolia do
   def wait(response = {:ok, _}), do: wait(response, 1000)
   def wait(response = {:error, _}), do: response
   def wait(response), do: response
+
+  @doc """
+  Push events to the Insights REST API.
+  Corresponds to https://www.algolia.com/doc/rest-api/insights/#push-events
+  `events` should be a List of Maps, each Map having the fields described in the link above
+  """
+  def push_events(events), do: push_events(events, 0)
+
+  defp push_events(_events, 4) do
+    {:error, "Unable to connect to Algolia Insights"}
+  end
+
+  defp push_events(events, curr_retry) do
+    :post
+    |> :hackney.request(
+      "https://insights.algolia.io/1/events",
+      request_headers([]),
+      Jason.encode!(%{"events" => events}),
+      [
+        :with_body,
+        path_encode_fun: &URI.encode/1,
+        connect_timeout: 3_000 * (curr_retry + 1),
+        recv_timeout: 30_000 * (curr_retry + 1),
+        ssl_options: [{:versions, [:"tlsv1.2"]}]
+      ]
+    )
+    |> case do
+      {:ok, code, _headers, response} when code in 200..299 ->
+        {:ok, Jason.decode!(response)}
+
+      {:ok, code, _, response} ->
+        {:error, code, response}
+
+      _ ->
+        push_events(events, curr_retry + 1)
```

Comment:
This implement was a bit odd (hard coding many options, code in 200...299 vs two comparisons, why retries) but looks like it was copied/pasted from defp send_request. Copy/paste is ofc not good for maintainability especially since the underlying code is not intuitive. I think all that might be needed is overloading `request_url` actually and renaming the `read_or_write` variable of send_request.

Reply:
i've re-implemented it according to the using `send_request` approach. could you take another look?